### PR TITLE
common: MemoryModel: performance improvements and API changes

### PR DESCRIPTION
**Pull request by ronen-fr**
**Performance:** per my off-line (code snippets) benchmarks
(see [benchtest](https://github.com/ronen-fr/memorymodel)):

- the parsing of /proc/status improved by 
1. only opening the proc file once (saves almost 40%)
2. using charconv

- for /proc/maps:
1. opening once (most of the improvement);
2. faster discarding of lines;
3. using charconv

**The API changes:**
- no output parameters;
- no internally-maintained samples;
- not issuing warning messages to the log directly;

The PR includes work by @athanatos & @shreekarSS,
picked from https://github.com/ceph/ceph/pull/54698

### CODE REVIEW

**Comment by ronen-fr** [comment]
jenkins test make check

**Comment by ronen-fr** [comment]
@vshankar - can you review? thanks

**Comment by vshankar** [comment]
> @vshankar - can you review? thanks

Ofcourse 👍 

(I will plan to have a look in the next few days)

File: src/common/MemoryModel.cc [hunk 1]
```diff
@@ -72,35 +99,47 @@ std::optional<int64_t> MemoryModel::get_mapped_heap()
...
+
+  while (!proc_status.eof() && yet_to_find > 0) {
+    string ln;
+    getline(proc_status, ln);
+
+    if (cmp_against(ln, "VmSize:", s.size) ||
+	cmp_against(ln, "VmRSS:", s.rss) || cmp_against(ln, "VmHWM:", s.hwm) ||
+	cmp_against(ln, "VmLib:", s.lib) ||
+	cmp_against(ln, "VmPeak:", s.peak) ||
+	cmp_against(ln, "VmData:", s.data)) {
```
Comment (mchangir) [inline src/common/MemoryModel.cc:132]: Would it be too much to push all required items in a hash table and test for the presence of the prefix of the current line in the hash table instead ?

**Review by mchangir — COMMENTED** [review]
just a small query towards improving performance

File: src/common/MemoryModel.cc (hunk 1 above)
Comment (ronen-fr) [reply src/common/MemoryModel.cc:132]: @mchangir ,
> Would it be too much to push all required items in a hash table and test for the presence of the prefix of the current line in the hash table instead ?

I do not see a way to do have better performance this way (I had originally measured multiple ideas I could think of). For example - note that we have sub-strings of at least two lengths here.

**Review by mchangir — APPROVED** [review]
LGTM

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/67020](https://tracker.ceph.com/issues/67020).

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20240801064407-debug

fs approved.

**Pull request merged**